blocks/api/serializer.js

```diff
@@ -37,34 +37,45 @@ export function getSaveContent( save, attributes ) {
 }
 
 /**
- * Returns comment attributes as serialized string, determined by subset of
- * difference between actual attributes of a block and those expected based
- * on its settings.
+ * Returns attributes which ought to be saved and serialized
  *
- * @param  {Object} realAttributes     Actual block attributes
- * @param  {Object} expectedAttributes Expected block attributes
- * @return {string}                    Comment attributes
+ * When a block exists in memory it contains as its attributes
+ * both those which come from the block comment opening _and_
+ * those which come from parsing the contents of the block.
+ * Additionally they may live in a form fine for memory but
+ * which isn't valid for serialization.
+ *
+ * This function returns a filtered set of attributes which are
+ * the ones which need to be saved in order to ensure a full
+ * serialization and they are in a form which is safe to store.
+ *
+ * @param {Object<String,*>}   allAttributes Attributes from in-memory block data
+ * @param {Object<String,*>}   fromContent   Attributes which are inferred from block content
+ * @returns {Object<String,*>} filtered set of attributes for minimum safe save/serialization
  */
-export function getCommentAttributes( realAttributes, expectedAttributes ) {
-	// Find difference and build into object subset of attributes.
-	const keys = difference(
-		Object.keys( realAttributes ),
-		Object.keys( expectedAttributes )
-	);
+export function attributesToSave( allAttributes, fromContent ) {
+	// Reasons an attribute need not be saved
+	const canBeInferred = key => fromContent.hasOwnProperty( key );
+	const isUndefined = key => undefined === allAttributes[ key ];
 
-	// Serialize the comment attributes as `key="value"`.
-	return keys.reduce( ( memo, key ) => {
-		const value = realAttributes[ key ];
-		if ( undefined === value ) {
-			return memo;
-		}
+	const isValid = key => ! some( [
+		isUndefined,
+		canBeInferred,
+	], f => f( key ) );
 
-		if ( 'string' === typeof value ) {
-			return memo + `${ key }="${ value.replace( '"', '\"' ) }" `;
-		}
+	// Specific ways we need to transform the values of saved attributes
+	const escapeDoubleQuotes = value => 'string' === typeof value
+		? value.replace( '"', '\"' )
+		: value;
 
-		return memo + `${ key }="${ value }" `;
-	}, '' );
+	const transform = key => escapeDoubleQuotes( allAttributes[ key ] );
+
+	// Iterate over attributes and produce the set to save
+	return reduce(
+		Object.keys( allAttributes ),
+		( toSave, key ) => Object.assign( toSave, isValid( key ) && { [ key ]: transform( key ) } ),
```

Comment:
> This feels a bit too magical.

I can see where it feels magical, but it's just the behavior of `Object.assign()` [by specification](http://www.ecma-international.org/ecma-262/6.0/#sec-object.assign).

> how about filter

this seems very reasonable. I'll update that.

> Also, do we really need isValid at all? Isn't it enough to say simply !! attributesFromContent[ key ] ?

well if that's all we said we'd be wrong in the case where `allAttributes[ key ]` is `undefined` _and_ it doesn't exist in `attributesFromContent`. trying to pull out the actual operations that are happening to determine if a key is valid is part of why I created `isValid()` since before it was spread across different lines.